### ISSUE TITLE
Revisit the output flagging (on exit) protocol, implementation & docs

### ISSUE THREAD

**Issue by pdobacz** [omisego/plasma-contracts#470]
## Issue Type 

```
[x] bug report
[ ] feature request
```

## Current Behavior

consider the following scenario:
1. Alice and Bob sign `tx1`, spending outputs `oA` and `oB` respectively, both "old" outputs
2. Bob signs `tx2`, spending `oB`
3. `tx1` is submitted and withheld
4. `tx2` is submitted, included and withheld
4. Bob IFEs with `tx2` (first!)
5. Alice IFEs with `tx1`
6. Both IFEs get challenged, but `tx2` IFE gets a response and is canonical, `tx1` is non canonical
7. Bob's IFE finalizes 1st and flags Bob's input, Bob gets oB out
8. Due to the flag rule, Alice's IFE will not finalize, because oB is flagged, leaving her without `oA`

https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/docs/design/payment-game-implementation-v1.md was used as source, which is a summary of current impl.

## Expected Behavior

Alice should be able to exit `oA`

## Suggested Fix

Something was already proposed by @paulperegud, pls do the honors

As part of the effort I suggest, that precise specs of this part of the protocol is produced somewhere in a suitable spot. Then pls combine this with fixing https://github.com/omisego/elixir-omg/blob/master/docs/standard_vs_in_flight_exits_interaction.md, per this comment: https://github.com/omisego/elixir-omg/pull/1118#issuecomment-557132890

> docs/specs should hold the current design of the mechanics and (if necessary) explanation why it works, in its entirety. Just state what and when is flagged and when flags are checked and that's it.
>
> All the links to prior discussions/PRs should only be used to provide historical background. But one should not need to follow those links to understand how and why things work.

## Motivation for Change

ask Alice :)

**Comment by boolafish** [comment]
Note from slack

by @paulperegud 
```
1) flag only when finalizing the exit
2) for canonical mark all inputs spend and exited outputs spend
3) for non-canonical mark only exiting inputs
```

However, above would not solve the exact scenario as stated in the issue. Thus second patch:
```
Looks like the problem is skipping IFE when input flagged? Probably should treat it as non canonical instead?
```

**Comment by boolafish** [comment]
seems like the proposal of treat input flagged as non-canonical would also solve the original issue: https://github.com/omisego/plasma-contracts/issues/102#issuecomment-495809967

So even in the challenge canonicity game it end up being both tx1 and tx2 canonical, while process IFE, assumes tx1 comes first, then tx1 would be canonical exiting output and tx2 non-canonical and exiting inputs. However, the shared input would be piggyback challenged so the invalid one would not withdraw. Seems working

**Comment by boolafish** [comment]
code fix for this (#483) merged, still need to enhance the elixir doc also planning to add some e2e IFE test

**Comment by boolafish** [comment]
All PRs (including docs) merged. closing this issue.
